[release-4.11] OCPBUGS-3560: Allow PROXY protocol for "Private" strategy

Miciah opened the pull request:
This is a manual cherry-pick of #803.  The changes include an openshift/api bump.

---

#### Bump openshift/api for PROXY protocol for Private

Bump to github.com/openshift/api@398424d53f7473dcc8be52f78692d2bea9b46b2d to get the "Protocol" field in the "Private" endpoint publishing strategy parameters to allow configuring PROXY protocol.

* `go.mod`: Update.
* `go.sum`:
* `manifests/00-custom-resource-definition.yaml`:
* `pkg/manifests/bindata.go`:
* `vendor/github.com/openshift/api/*`:
* `vendor/modules.txt`: Regenerate.


#### Allow PROXY protocol for "Private" strategy

Allow the user to configure an IngressController with the "Private" endpoint publishing strategy type to use PROXY protocol.

* `pkg/operator/controller/ingress/controller.go` (`setDefaultPublishingStrategy`): Check whether the user specified the protocol on the "Private" endpoint publishing strategy parameters, and update status as necessary.
(`IsProxyProtocolNeeded`): Check whether the user specified the protocol on the "Private" endpoint publishing strategy parameters.
* `pkg/operator/controller/ingress/controller_test.go` (`TestSetDefaultPublishingStrategyHandlesUpdates`, `TestIsProxyProtocolNeeded`): Add test cases for the "Private" endpoint publishing strategy.

candita commented:
/assign @gcs278

ShudiLi commented:
From QE side, tested it with 4.11.0-0.ci.test-2023-04-27-005005-ci-ln-hbmijl2-latest
1.
% oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.11.0-0.ci.test-2023-04-27-005005-ci-ln-hbmijl2-latest   True        False         19m     Cluster version is 4.11.0-0.ci.test-2023-04-27-005005-ci-ln-hbmijl2-latest
%

2. check the status.endpointPublishingStrategy, expect enable the proxy: {"private":{"protocol":"PROXY"},"type":"Private"}
% oc -n openshift-ingress-operator get ingresscontrollers/int1 -o 'jsonpath={.status.endpointPublishingStrategy}'
{"private":{"protocol":"PROXY"},"type":"Private"}
%            

3. check the env ROUTER_USE_PROXY_PROTOCOL, expect {"name":"ROUTER_USE_PROXY_PROTOCOL","value":"true"}
% oc -n openshift-ingress get deployments/router-int1 -o 'jsonpath={.spec.template.spec.containers[0].env[?(@.name=="ROUTER_USE_PROXY_PROTOCOL")]}'
{"name":"ROUTER_USE_PROXY_PROTOCOL","value":"true"}%                                                                                                                                                                                                                      %                                 

/label qe-approved
thanks                                                                                                                                                                             %

gcs278 commented on pkg/manifests/bindata.go:
```diff
@@ -946,11 +946,13 @@ var _bindata = map[string]func() (*asset, error){
-//       img/
-//         a.png
-//         b.png
+//
```
This is why ci/prow/verify is failing. It was spaces and now is indents. Is this a go version issue as you mentioned here: https://github.com/openshift/cluster-ingress-operator/pull/869#issuecomment-1351768652?

Miciah replied:
Sorry!  Fixed in https://github.com/openshift/cluster-ingress-operator/compare/ba9df6ba6efecdb5b40eabb06cb9dfe20df36d75..be7ae90c7363152360fc01a913def9c21d49dc62.

gcs278 commented:
I'm the only reviewer so
/approve
/lgtm

Miciah commented:
[verify failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/914/pull-ci-openshift-cluster-ingress-operator-release-4.11-verify/1651685368900096000) because I had forgotten to `git add vendor/github.com/openshift/api/models-schema` after `go mod vendor`.  https://github.com/openshift/cluster-ingress-operator/compare/be7ae90c7363152360fc01a913def9c21d49dc62..2fc2f3b16f9777160972b5b23bc51cd3384cd518 adds the missing file.

Miciah commented:
[e2e-aws-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/914/pull-ci-openshift-cluster-ingress-operator-release-4.11-e2e-aws-operator/1651688541622636544) because `TestHostNetworkPortBinding` failed:
```
=== RUN   TestAll/serial/TestHostNetworkPortBinding
=== PAUSE TestAll/serial/TestHostNetworkPortBinding
=== CONT  TestAll/serial/TestHostNetworkPortBinding
    operator_test.go:1009: Expected conditions: map[Admitted:True Available:True DNSManaged:False DeploymentReplicasAllAvailable:True LoadBalancerManaged:False]
         Current conditions: map[Admitted:True Available:True DNSManaged:False Degraded:False DeploymentAvailable:True DeploymentReplicasAllAvailable:False DeploymentReplicasMinAvailable:True LoadBalancerManaged:False PodsScheduled:False Progressing:False Upgradeable:True]
    operator_test.go:1011: failed to observe expected conditions for the second ingresscontroller: timed out waiting for the condition
    operator_test.go:1034: deleted ingresscontroller samehost
    operator_test.go:1034: deleted ingresscontroller hostnetworkportbinding
```
From [`events.json`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/914/pull-ci-openshift-cluster-ingress-operator-release-4.11-e2e-aws-operator/1651688541622636544/artifacts/e2e-aws-operator/gather-extra/artifacts/events.json), it seems that the "samehost" router pod got deleted and failed to be rescheduled:
```
% jq -rc < events.json '.items|sort_by(.metadata.creationTimestamp)|.[]|select(.involvedObject.name//""|contains("samehost"))|.metadata.creationTimestamp+" ("+(.count//1|tostring)+"x) "+.reason+": "+.message' 
2023-04-27T21:36:48Z (1x) FailedToCreateEndpoint: Failed to create endpoint for service openshift-ingress/router-internal-samehost: endpoints "router-internal-samehost" already exists
2023-04-27T21:36:48Z (1x) Admitted: ingresscontroller passed validation
2023-04-27T21:36:48Z (2x) CreatedDefaultCertificate: Created default wildcard certificate "router-certs-samehost"
2023-04-27T21:36:48Z (1x) Scheduled: Successfully assigned openshift-ingress/router-samehost-54dbc7dd5f-9vxhj to ip-10-0-136-46.us-east-2.compute.internal by ip-10-0-228-168
2023-04-27T21:36:48Z (1x) FailedMount: MountVolume.SetUp failed for volume "metrics-certs" : secret "router-metrics-certs-samehost" not found
2023-04-27T21:36:48Z (1x) FailedMount: MountVolume.SetUp failed for volume "default-certificate" : secret "router-certs-samehost" not found
2023-04-27T21:36:48Z (1x) SuccessfulCreate: Created pod: router-samehost-54dbc7dd5f-9vxhj
2023-04-27T21:36:48Z (1x) ScalingReplicaSet: Scaled up replica set router-samehost-54dbc7dd5f to 1
2023-04-27T21:36:49Z (1x) Pulled: Container image "registry.build01.ci.openshift.org/ci-op-bibmbvxp/stable@sha256:b477ca8dbeaa1c77f6dfa3af5e8eaf36cf25ae8b9acc2ccc1cc48bf2c89077f7" already present on machine
2023-04-27T21:36:49Z (1x) Created: Created container router
2023-04-27T21:36:49Z (1x) Started: Started container router
2023-04-27T21:37:18Z (1x) Killing: Stopping container router
2023-04-27T21:37:18Z (1x) FailedScheduling: 0/6 nodes are available: 1 node(s) were unschedulable, 2 node(s) didn't match Pod's node affinity/selector, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling.
2023-04-27T21:37:18Z (1x) SuccessfulCreate: Created pod: router-samehost-54dbc7dd5f-rtsp6
2023-04-27T21:37:23Z (5x) ProbeError: Readiness probe error: HTTP probe failed with statuscode: 500
body: [+]backend-http ok
[+]has-synced ok
[-]process-running failed: reason withheld
healthz check failed


2023-04-27T21:37:23Z (5x) Unhealthy: Readiness probe failed: HTTP probe failed with statuscode: 500
2023-04-27T21:39:39Z (1x) FailedScheduling: 0/6 nodes are available: 2 node(s) didn't match Pod's node affinity/selector, 2 node(s) had untolerated taint {node-role.kubernetes.io/master: }, 2 node(s) were unschedulable. preemption: 0/6 nodes are available: 6 Preemption is not helpful for scheduling.
2023-04-27T21:41:48Z (3x) DeletedDeployment: Deleted deployment openshift-ingress/router-samehost
2023-04-27T21:41:48Z (1x) FailedScheduling: skip schedule deleting pod: openshift-ingress/router-samehost-54dbc7dd5f-rtsp6
```
It looks like nodes are getting rebooted while the tests are running, which could explain why the "samehost" router pod was deleted and couldn't be rescheduled:
```
%  jq -rc < events.json '.items|sort_by(.metadata.creationTimestamp)|.[]|select(.involvedObject.kind=="Node")|.metadata.creationTimestamp+" ("+(.count//1|tostring)+"x) "+.involvedObject.name+" "+.reason+": "+.message'                                                                                                                                                                                                                
[...]
2023-04-27T21:36:56Z (1x) ip-10-0-178-19.us-east-2.compute.internal Rebooted: Node ip-10-0-178-19.us-east-2.compute.internal has been rebooted, boot id: d8e11df6-1669-43f6-96cd-cb0fd63b0890                                                                                                                                                                                                                                           
2023-04-27T21:36:56Z (1x) ip-10-0-178-19.us-east-2.compute.internal NodeReady: Node ip-10-0-178-19.us-east-2.compute.internal status is now: NodeReady                                                                                                                                                                                                                                                                                  
2023-04-27T21:36:56Z (1x) ip-10-0-178-19.us-east-2.compute.internal NodeNotSchedulable: Node ip-10-0-178-19.us-east-2.compute.internal status is now: NodeNotSchedulable                                                                                                                                                                                                                                                                
2023-04-27T21:37:01Z (1x) ip-10-0-178-19.us-east-2.compute.internal Starting: openshift-sdn done initializing node networking.                                                                                                                                                                                                                                                                                                          
2023-04-27T21:37:05Z (1x) ip-10-0-182-216.us-east-2.compute.internal Starting: openshift-sdn done initializing node networking.                                                                                                                                                                                                                                                                                                         
2023-04-27T21:37:06Z (1x) ip-10-0-182-216.us-east-2.compute.internal FastControllerResync: Controller "CheckEndpointsTimeToStart" resync interval is set to 0s which might lead to client request throttling                                                                                                                                                                                                                            
2023-04-27T21:37:06Z (1x) ip-10-0-182-216.us-east-2.compute.internal FastControllerResync: Controller "CheckEndpointsStop" resync interval is set to 0s which might lead to client request throttling                                                                                                                                                                                                                                   
2023-04-27T21:37:12Z (1x) ip-10-0-178-19.us-east-2.compute.internal Uncordon: Update completed for config rendered-worker-7b7e40488c1dad43017388e8822faf7f and node has been uncordoned                                                                                                                                                                                                                                                 
2023-04-27T21:37:12Z (1x) ip-10-0-178-19.us-east-2.compute.internal NodeDone: Setting node ip-10-0-178-19.us-east-2.compute.internal, currentConfig rendered-worker-7b7e40488c1dad43017388e8822faf7f to Done                                                                                                                                                                                                                            
2023-04-27T21:37:12Z (1x) ip-10-0-178-19.us-east-2.compute.internal ConfigDriftMonitorStarted: Config Drift Monitor started, watching against rendered-worker-7b7e40488c1dad43017388e8822faf7f                                                                                                                                                                                                                                          
2023-04-27T21:37:13Z (1x) ip-10-0-136-46.us-east-2.compute.internal ConfigDriftMonitorStopped: Config Drift Monitor stopped                                                                                                                                                                                                                                                                                                             
2023-04-27T21:37:13Z (1x) ip-10-0-136-46.us-east-2.compute.internal Cordon: Cordoned node to apply update                                                                                                                                                                                                                                                                                                                               
2023-04-27T21:37:13Z (1x) ip-10-0-136-46.us-east-2.compute.internal Drain: Draining node to update config.                                                                                                                                                                                                                                                                                                                              
2023-04-27T21:37:16Z (1x) ip-10-0-178-19.us-east-2.compute.internal NodeSchedulable: Node ip-10-0-178-19.us-east-2.compute.internal status is now: NodeSchedulable                                                                                                                                                                                                                                                                      
2023-04-27T21:37:18Z (1x) ip-10-0-182-216.us-east-2.compute.internal Uncordon: Update completed for config rendered-master-8a8418bdac10dea73efedbf2a4276f74 and node has been uncordoned                                                                                                                                                                                                                                                
2023-04-27T21:37:18Z (1x) ip-10-0-182-216.us-east-2.compute.internal NodeDone: Setting node ip-10-0-182-216.us-east-2.compute.internal, currentConfig rendered-master-8a8418bdac10dea73efedbf2a4276f74 to Done                                                                                                                                                                                                                          
[...]
```
I wouldn't expect nodes to reboot during E2E tests, but anyway, it doesn't look like a problem with cluster-ingress-operator or with the test itself.  
/test e2e-aws-operator

Miciah commented:
[e2e-aws-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/914/pull-ci-openshift-cluster-ingress-operator-release-4.11-e2e-aws-operator/1651744040179208192)  because the kube-apiserver, kube-controller-manager, kube-scheduler, and network clusteroperators failed to roll out updates.  
/test e2e-aws-operator

gcs278 commented:
/lgtm

Miciah commented:
/label backport-risk-assessed

ShudiLi commented:
/label cherry-pick-approved
thanks
